### PR TITLE
Sf object tracking integration

### PR DESCRIPTION
**Pull request by SvenStruanFinlay**
Prioritise is all there for smart prio. Spent a bit trying to add more complex things but i wanna get report done first. I know there are gonna be some changes requested somewhere so leeme know when you can.

### CODE REVIEW

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/api_controller.cpp [hunk 1]
```diff
@@ -10,19 +10,17 @@ void ApiController::add_user(ApiUserInterface& user)
...
+void ApiController::new_items(const std::vector<ClassificationItem>& results)
 {
 	SPDLOG_INFO("Sending items to API users");
 
-	int time = std::clock();
+	auto time = std::clock();
 
 	SPDLOG_INFO("Testing Time counter, time is {} outside loop", std::to_string(time - duration));
-	if ((time - duration) > 2000)
-	{
```
Comment (lspcool) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/api_controller.cpp:21]: I want to know what is the initial purpose for (time - duration) > 2000)

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_clock.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+{
+	SPDLOG_INFO("Item {}, inserting into timer map",id);
+	auto F = cooldowns.find(id);
+	if (F == cooldowns.end())
+
+	{
+		SPDLOG_INFO("Item {}, inserting into timer map sucessful", id);
+		auto time = clock();
+		cooldowns.insert(std::pair<int, clock_t>(id, time));
+		return true;
```
Comment (lspcool) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_clock.cpp:69]: Maybe put LOG after insert

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_clock.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
+#include "priority_clock.hpp"
+
+
+bool priority_clock::check_cooldown(int id, int time_dif)
+{
+	auto time = clock();
+	auto F = cooldowns.find(id);
```
Comment (lspcool) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_clock.cpp:14]: the id is not unique, the (object name+id) is unique.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/output_stream_controller.cpp [hunk 4]
```diff
@@ -41,72 +41,88 @@ void OutputStreamController::output_to_depth_window(const rs2::frame& frame) con
...
+	/// <summary>
+	/// Tells the system if the opencv stream output windows are ready, if enabled.
+	/// In the case where no output stream windows are enabled, returns true.
+	/// Based on the rs-dnn and rs-imshow examples.
+	/// </summary>
+	/// <returns>
+	///   Returns TRUE if the windows can receive frames, or if no output stream windows are enabled.
+	///   Returns FALSE if the windows can not receive frames.
+	/// </returns>
+
```
Comment (lspcool) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/output_stream_controller.cpp]: The summary exists in the hpp file

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/api_controller.cpp (hunk 1 above)
Comment (SvenStruanFinlay) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/api_controller.cpp:21]: So this was before the colldown system was implemented and it basically stoped the service from sending data every frame. sent it two seconds instead

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_clock.cpp (hunk 2 above)
Comment (SvenStruanFinlay) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_clock.cpp:69]: On it

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/output_stream_controller.cpp (hunk 4 above)
Comment (SvenStruanFinlay) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/output_stream_controller.cpp]: fair

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/CMakeLists.txt [hunk 5]
```diff
@@ -35,6 +35,9 @@ set(SRC_FILES
     ${PROJECT_SOURCE_DIR}/Intel_SighPP_CMake/src/priority_lib/priority_module.cpp
     ${PROJECT_SOURCE_DIR}/Intel_SighPP_CMake/src/priority_lib/size_priority.cpp
     ${PROJECT_SOURCE_DIR}/Intel_SighPP_CMake/src/priority_lib/priority.hpp
+    ${PROJECT_SOURCE_DIR}/Intel_SighPP_CMake/src/priority_lib/priority_clock.hpp
+    ${PROJECT_SOURCE_DIR}/Intel_SighPP_CMake/src/priority_lib/priority_clock.cpp
+    ${PROJECT_SOURCE_DIR}/Intel_SighPP_CMake/src/priority_lib/priority_clock.hpp
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/CMakeLists.txt:39]: Isn't this a duplicate of line 38?

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/api_controller.hpp [hunk 6]
```diff
@@ -19,10 +19,6 @@ class ApiController
 	/// </summary>
 	/// <param name="user">User to add to current users.</param>
 	void add_user(ApiUserInterface& user);
+	void new_items(const std::vector<ClassificationItem>& results);
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/api_controller.hpp:27]: Was there a reason for removing the documentation of this method?

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/output_stream_controller.cpp [hunk 7]
```diff
@@ -41,72 +41,88 @@ void OutputStreamController::output_to_depth_window(const rs2::frame& frame) con
...
+	bool OutputStreamController::should_receive_new_frames() const
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/output_stream_controller.cpp]: Why are these methods suddenly indented?

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/output_stream_controller.cpp (hunk 4 above)
Comment (GisliGeorgs) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/output_stream_controller.cpp]: Yeah, it should be in the header file, not the source file 😄

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_clock.cpp [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	id.append(std::to_string(item.id));
+	return id;
+}
+
+bool priority_clock::check_cooldown(std::string id, int time_dif)
+{
+	auto F = cooldowns.find(id);
+	auto time = clock();
+
+	SPDLOG_INFO("Item Finding {}, looking for time {}", id, time_dif);
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_clock.cpp]: Wouldn't 'Finding Item {}' be more descriptive? 😄

Comment (GisliGeorgs) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_clock.cpp]: And in the next SPDLOG statement 😀

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_clock.cpp [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+		auto time = clock();
+		cooldowns.insert(std::pair<std::string, clock_t>(id, time));
+
+		SPDLOG_INFO("Item {}, inserting into timer map sucessful", id);
+		return true;
+	}
+	return false;
+}
+
+bool priority_clock::remove_new_id(std::string id)
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_clock.cpp:74]: Remove New Id? Is it not just Remove Id?

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_clock.hpp [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+	/// <returns> True if  clock - value > time_dif else False</returns>
+	std::string get_unique_id(ClassificationItem& item);
+	bool check_cooldown(std::string, int time_difid);
+	bool check_contains(std::string id);
+	bool update_timer(std::string id);
+	bool insert_new_id(std::string id);
+	bool remove_new_id(std::string id);
+	bool clear_ids(int time_dif);
+
+
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_clock.hpp:36]: Weird empty lines 😄 and above.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_module.cpp [hunk 11]
```diff
@@ -4,6 +4,35 @@
 #include "priority_module.hpp"
 int pivot = 0;
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_module.cpp]: I know its not part of this commit, but shouldn't this variable be part of the priority_module?
Instead of being a global variable in this file?

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/smart_priority.cpp [hunk 12]
```diff
@@ -3,15 +3,47 @@
...
-	msg_add_name(item);
-	msg_add_distance(item);
+	if (item.track_point < 5)
+	{
+		item.priority = Priority::LOW;
+		return;
+	}
+
+
+	SPDLOG_INFO("/n/n Processing item {} : {}", item.name, item.id);
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/smart_priority.cpp:37]: Why Empty lines?

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/smart_priority.hpp [hunk 13]
```diff
@@ -1,76 +1,155 @@
...
+	int bottom = 0;
 	int mid_w = 0;
+	priority_clock pc;
+	Priority min = Priority::UNDEFINED;
 
 public:
 
+	/// <summary>
+	///Object Creation. Not to be inherited furhter.
+	/// Initialises location markers based on ratio
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/smart_priority.hpp]: based on what ratio? 😄

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/smart_priority.hpp [hunk 14]
```diff
@@ -1,76 +1,155 @@
...
+	/// <summary>
+	/// Determines location markers for objects within the screen. 
+	/// </summary>
 	void determine_location_markers();
+
+
+	//________data and message control________: 
+
+	/// <summary>
+	/// Based on rectangle of object determin location
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/smart_priority.hpp]: Is this good english? 😄 
Also, if the item param is not documented, maybe include something about it in the summary 😀

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/smart_priority.hpp [hunk 15]
```diff
@@ -1,76 +1,155 @@
...
+	double time_until_colision(ClassificationItem& item);
+
+	//__________Tracking checking rules______________
+
+	/// <summary>
+	/// Tracks the main cooldown of each object. If item is out of cooldown then it 
+	/// is approved to run through the priority rules. If not it is currently 
+	/// set to low prio.
+	///
+	/// Note: Items that are tracked but not currently seen are also set to low prio. This could be improved after report.
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/smart_priority.hpp]: Maybe better to include the last part as a // TODO, instead of a documented note 😄

**Review by GisliGeorgs — CHANGES_REQUESTED** [review]
Look at comments 😄 Nothing major.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/Intel_Sight_++.cpp [hunk 16]
```diff
@@ -206,6 +207,7 @@ int main(int argc, char** argv)
 
 	auto profile = config.get_stream(RS2_STREAM_COLOR).as<rs2::video_stream_profile>();
 
+
```
Comment (XingdaCen) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/Intel_Sight_++.cpp]: Can remove this empty line.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/output_stream_controller.cpp [hunk 17]
```diff
@@ -41,72 +41,88 @@ void OutputStreamController::output_to_depth_window(const rs2::frame& frame) con
...
-
-void OutputStreamController::output_to_color_window(cv::Mat color_matrix, cv::Mat depth_matrix, const std::vector<PrioritisedClassificationResult>& vector) const
+void OutputStreamController::output_to_color_window(cv::Mat color_matrix, cv::Mat depth_matrix, const std::vector<ClassificationItem>& vector) const
 {
-	for (auto&& prioritised_classification_result : vector)
+
+	//const auto width = frame.as<rs2::video_frame>().get_width();
+	//const auto height = frame.as<rs2::video_frame>().get_height();
+
+	//const cv::Mat color_mat(cv::Size(width, height), CV_8UC3, const_cast<void*>(frame.get_data()), cv::Mat::AUTO_STEP);
```
Comment (XingdaCen) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/output_stream_controller.cpp:50]: Do we need to use the three lines that are commented out?

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/ml_lib/caffe_impl.cpp [hunk 18]
```diff
@@ -26,7 +26,7 @@ struct CaffeModelImpl : public ModelInterface {
 	const float inScaleFactor = 0.007843f;
 	const float meanVal = 127.5;
 
-	const float confidence_threshold = 0.9f;
+	const float confidence_threshold = 0.7f;
```
Comment (XingdaCen) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/ml_lib/caffe_impl.cpp]: No problem here but can you change `opencv2\imgproc.hpp` to `opencv2/imgproc.hpp` in line 7?

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/api_controller.hpp (hunk 6 above)
Comment (SvenStruanFinlay) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/api_controller.hpp:27]: it may have disappeared in the merge

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/CMakeLists.txt (hunk 5 above)
Comment (SvenStruanFinlay) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/CMakeLists.txt:39]: Fixed.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/api_controller.hpp (hunk 6 above)
Comment (SvenStruanFinlay) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/api_controller.hpp:27]: Fixed.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/output_stream_controller.cpp (hunk 7 above)
Comment (SvenStruanFinlay) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/output_stream_controller.cpp]: Fixed.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_clock.cpp (hunk 8 above)
Comment (SvenStruanFinlay) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_clock.cpp]: Fixed.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_module.cpp (hunk 11 above)
Comment (SvenStruanFinlay) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/priority_module.cpp]: Fixed.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/smart_priority.cpp (hunk 12 above)
Comment (SvenStruanFinlay) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/smart_priority.cpp:37]: Fixed.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/smart_priority.hpp (hunk 13 above)
Comment (SvenStruanFinlay) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/smart_priority.hpp]: Fixed.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/smart_priority.hpp (hunk 14 above)
Comment (SvenStruanFinlay) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/priority_lib/smart_priority.hpp]: Fixed.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/output_stream_controller.cpp (hunk 17 above)
Comment (SvenStruanFinlay) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/interface_controller/output_stream_controller.cpp:50]: Fixed.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/ml_lib/caffe_impl.cpp (hunk 18 above)
Comment (SvenStruanFinlay) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/ml_lib/caffe_impl.cpp]: Fixed.

**Review by XingdaCen — APPROVED** [review]

**Review by lspcool — APPROVED** [review]
